### DISCUSSION TITLE
Promiscuous mode

### DISCUSSION THREAD

**Discussion by Malvineous** [troglobit/smcroute#174] [Q&A] [answered]
I searched the README for "promiscuous" and found nothing so hopefully I haven't missed something obvious again :)

I noticed when setting up smcroute that no traffic was being forwarded at all.  However when I ran `tcpdump` on the same machine, looking at the source network interface, suddenly smcroute saw traffic, set up routes and everything was forwarded.  But as soon as I terminated `tcpdump` all multicast traffic stopped being forwarded.

After some experimentation, the issue turned out to be that the network interface was not in promiscuous mode.  Running tcpdump puts it into promiscuous mode which is why it was causing the incoming multicast packets to suddenly arrive on the machine.

Going into the OpenWRT settings and setting the source interface to be run permanently in promiscuous mode fixed the problem, and now packets are forwarded even when tcpdump is not running.

I am wondering whether this is something that smcroute should handle?  Should it put the source network interfaces into promiscuous mode automatically, or is that deemed to be the responsibility of the OS?  In which case would it be worth adding something to the README if this is a requirement for having smcroute function correctly?

**Comment by troglobit** [comment, marked as answer]
It's not really up to SMCRoute to handle the interface flags. If you want to set the interface in promiscuous mode you can do that. However, that means that the CPU will have to receive and drop every packet that is *not* supposed to be routed.  Depending on your hardware there are ways to filter multicast, e.g. IGMP/MLD snooping, which allows only "subscribed" (or joined) multicast groups to be forwarded to the CPU.  SMCRoute supports joining multicast groups, both ASM and SSM, meaning you can select exactly which groups reach the CPU and are forwarded by the multicast route.  See the `mgroup` setting in the documentation.

Moving this to Discussions, since it's more of a support question than bug/feature request.

**Comment by Malvineous** [comment]
Ah I see.  Evidently the switch hardware in my router (or OpenWRT) doesn't support this for IPv6, but it explains what the IGMP snooping options are for as I was wondering how they applied to a network interface.

Thanks for the explanation.  I see now that `mgroup` should be all SMCRoute needs to do and it's up to the network driver essentially to handle it from there.  I did try `mgroup` before posting my message but it didn't change anything.  I take your explanation to mean that would be the solution instead of promiscuous mode, if the router hardware and driver supported MLD snooping.

**Reply by troglobit** [reply]
Exactly.  Sometimes, as a workaround, switches have a "multicast router port" setting that allows forwarding (known and unknown) multicast through a port, e.g. the CPU port.  This allows forwarding of only multicast, which is better than promisc which forwards *everything*.  

The underlying Linux kernel, with its bridge module, is slowly becoming better and better at supporting multicast filtering/snooping, but it depends a lot on the chipset driver for the switching silicon in your router. The latest OpenWRT major release has also adopted using the kernel switchdev layer, instead of proprietary blobs and such, so support will gradually become better.
